// next-i18next/dist/commonjs/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "consoleMessage", {
  enumerable: true,
  get: function () {
    return _consoleMessage.default;
  }
});
Object.defineProperty(exports, "forceTrailingSlash", {
  enumerable: true,
  get: function () {
    return _forceTrailingSlash.default;
  }
});
Object.defineProperty(exports, "lngFromReq", {
  enumerable: true,
  get: function () {
    return _lngFromReq.default;
  }
});
Object.defineProperty(exports, "lngPathCorrector", {
  enumerable: true,
  get: function () {
    return _lngPathCorrector.default;
  }
});
Object.defineProperty(exports, "lngPathDetector", {
  enumerable: true,
  get: function () {
    return _lngPathDetector.default;
  }
});
Object.defineProperty(exports, "localeSubpathRequired", {
  enumerable: true,
  get: function () {
    return _localeSubpathRequired.default;
  }
});
Object.defineProperty(exports, "parseHref", {
  enumerable: true,
  get: function () {
    return _parseHref.default;
  }
});
Object.defineProperty(exports, "redirectWithoutCache", {
  enumerable: true,
  get: function () {
    return _redirectWithoutCache.default;
  }
});

var _consoleMessage = _interopRequireDefault(require("./console-message.js"));

var _forceTrailingSlash = _interopRequireDefault(require("./force-trailing-slash"));

var _lngFromReq = _interopRequireDefault(require("./lng-from-req"));

var _lngPathCorrector = _interopRequireDefault(require("./lng-path-corrector"));

var _lngPathDetector = _interopRequireDefault(require("./lng-path-detector"));

var _localeSubpathRequired = _interopRequireDefault(require("./locale-subpath-required"));

var _parseHref = _interopRequireDefault(require("./parse-href"));

var _redirectWithoutCache = _interopRequireDefault(require("./redirect-without-cache"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
